File: src/main/domain/algorithm/Algorithm.java
package domain.algorithm;

import domain.stock.History;
import utils.Pair;
import domain.stock.Holding;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import data.SerializationManager;

//===========================================================//
/**
 * Abstract base class for all trading algorithms.
 * <p> Defines the required interface and provides factory methods for initializing algorithms in different modes.</p>
 */
//===========================================================//

public abstract class Algorithm {
    //===========================================================//
    //===========================================================//
    // Public Interface(s)

    /**
     * Executes the algorithm based on current holdings and market conditions.
     *
     * @param holdings      - List of currently owned assets.
     * @param allocatedToke - Amount of capital allocated for trading.
     * @param currentPrice  - Current market price of the asset.
     * @return              AlgorithmOutput containing the decision/results.
     */
    public abstract AlgorithmOutput run(final List<Holding> holdings, final double allocatedToke, final double currentPrice);

    /**
     * Updates the internal state/history of the algorithm.
     *
     * @param stock - Historical stock data to be incorporated.
     */
    public abstract void updateHistory(final History stock);

    //===========================================================//
    /**
     * Initializes an algorithm instance configured for backtesting.
     *
     * @param type     - Type of algorithm to initialize.
     * @param stockNev - Stock identifier/name.
     * @param from     - Start date (inclusive).
     * @param to       - End date (inclusive).
     * @return         Pair containing the list of history that was not used up for initialisation and the algorithm instance.
     */
    static public final Pair<List<History>, Algorithm> initForBackTest(final Type type, final String stockNev, final int from, final int to) {
        return sm_initialiser(type, Init.BACKTEST, stockNev, from, to);
    }

    //===========================================================//

    /**
     * Initializes an algorithm instance configured for live trading.
     *
     * @param type     - Type of algorithm to initialize.
     * @param stockNev - Stock identifier/name.
     * @return         Initialized algorithm instance.
     */
    static public final Algorithm initForTrading(final Type type, final String stockNev) {
        return sm_initialiser(type, Init.TRADING, stockNev, Integer.MIN_VALUE, Integer.MAX_VALUE).second;
    }

    //===========================================================//
    //===========================================================//
    // Private Interface(s)

    /**
     * Core initialization method used by both trading and backtesting setups.
     *
     * @param type     - Algorithm type.
     * @param init     - Initialization mode.
     * @param stockNev - Stock identifier/name.
     * @param from     - Start date (inclusive).
     * @param to       - End date (inclusive).
     * @return         Pair of history data and initialized algorithm.
     */
    static private final Pair<List<History>, Algorithm> sm_initialiser(final Type type, final Init init, final String stockNev, final int from, final int to) {
        try {
            final var retHistory = sm_historyInitialiser(stockNev, from, to);
            final Algorithm retAlgorithm;

            switch (type) {
                case Type.TACPP46: {
                    retAlgorithm = new TACPP46(init, retHistory);
                } break;
                default: throw new IllegalArgumentException("Type");
            }

            return new Pair<>(retHistory, retAlgorithm);
        }
        catch(IOException IO_E) { throw new IllegalArgumentException(IO_E.getMessage(), IO_E.getCause()); }
    }

    //===========================================================//
    /**
     * Loads historical data from files based on the given parameters.
     *
     * @param stockNev - Stock identifier/name.
     * @param from     - Start date (inclusive).
     * @param to       - End date (inclusive).
     * @return         List of historical data entries.
     * 
     * @throws IOException If file reading fails.
     * @throws IllegalArgumentException If the expected range of files is not fully available.
     */

    static private final List<History> sm_historyInitialiser(final String stockNev, final int from, final int to) throws IOException {
        final var backtestFiles = new File("resources/backtest/us/").listFiles();
        final List<Pair<File, Integer>> proxy = new ArrayList<>();

        for(final var file : backtestFiles) {
            final String nameWithoutExtension = file.getName().replaceFirst("\\.json$", "");
            final int date = Integer.parseInt(nameWithoutExtension.substring(nameWithoutExtension.length() - 2));
            final String nameWithoutExtensionAndDate = nameWithoutExtension.substring(0, nameWithoutExtension.length() - 2);

            if(date >= from && date <= to && nameWithoutExtensionAndDate.equals(stockNev)) {
                proxy.add(new Pair<>(file, date));
            }
        }

        
        proxy.sort(Comparator.comparingInt(h -> h.second)); // Sort chronologically by the date
        // If 'from' and/or 'to' are MIN and/or MAX then we've loaded all, so we don't throw.
        if(from != Integer.MIN_VALUE && to != Integer.MAX_VALUE && proxy.size() != (to - from + 1)) throw new IllegalArgumentException("From-To");

        final List<History> ret = new ArrayList<>();
        for(final var item : proxy) {
            final var serData = SerializationManager.loadFromFile(item.first);
            for(final var item2 : serData.stockHistory.entrySet()) {
                ret.addAll(item2.getValue());
            }
        }

        return ret;
    }

    //===========================================================//
    //===========================================================//
    // Enum(s)

    /**
     * Supported algorithm types.
     */
    static public enum Type {
        TACPP46
    }

    //===========================================================//
    /**
     * Initialization modes for the algorithm.
     */
    static protected enum Init {
        BACKTEST,
        TRADING
    }

    //===========================================================//
    //===========================================================//
    // Constructor(s)

    /**
     * Protected constructor to enforce initialization mode handling
     * in subclasses.
     *
     * @param init - Initialization mode.
     */
    protected Algorithm(final Init init) {}
}

File: src/main/kotlin/domain/algorithm/Algorithm.kt
package domain.algorithm

import data.SerializationManager
import domain.stock.History
import domain.stock.Holding
import java.io.File

//===========================================================//
/**
 * Abstract base class for all trading algorithms.
 * 
 *  Defines the required interface and provides factory methods for initializing algorithms in different modes.
 */
//===========================================================//

abstract class Algorithm {
    //===========================================================//
    //===========================================================//
    // Public Method(es)
    /**
     * Executes the algorithm based on current holdings and market conditions.
     *
     * @param holdings      - List of currently owned assets.
     * @param allocatedCapital - Amount of capital allocated for trading.
     * @param currentPrice  - Current market price of the asset.
     * @return              AlgorithmOutput containing the decision/results.
     */
    abstract fun run(holdings: List<Holding>, allocatedCapital: Double, currentPrice: Double): Output

    //===========================================================//
    /**
     * Updates the internal state of the algorithm.
     *
     * @param history - Historical stock data to be incorporated.
     */
    abstract fun updateHistory(history: History)

    companion object {
        //===========================================================//
        /**
         * Initializes an algorithm instance configured for backtesting.
         *
         * @param type     - Type of algorithm to initialize.
         * @param stockName - Stock identifier/name.
         * @param from     - Start date (inclusive).
         * @param to       - End date (inclusive).
         * @return         Pair containing the list of history that was not used up for initialisation and the algorithm instance.
         */
        fun initForBackTest(type: Type, stockName: String, from: Int, to: Int): Pair<List<History>, Algorithm> {
            return initialiser(type, Algorithm.Init.BACKTEST, stockName, from, to)
        }

        //===========================================================//
        /**
         * Initializes an algorithm instance configured for live trading.
         *
         * @param type     - Type of algorithm to initialize.
         * @param stockName - Stock identifier/name.
         * @return         Initialized algorithm instance.
         */
        fun initForTrading(type: Algorithm.Type, stockName: String): Algorithm {
            return initialiser(type, Init.TRADING, stockName, Int.MIN_VALUE, Int.MAX_VALUE).second
        }

        //===========================================================//
        //===========================================================//
        // Private Method(es)
        /**
         * Core initialization method used by both trading and backtesting setups.
         *
         * @param type     - Algorithm type.
         * @param init     - Initialization mode.
         * @param stockName - Stock identifier/name.
         * @param from     - Start date (inclusive).
         * @param to       - End date (inclusive).
         * @return         Pair of history data and initialized algorithm.
         */
        private fun initialiser(type: Type, init: Init, stockName: String, from: Int, to: Int): Pair<List<History>, Algorithm> {
            val retHistory: List<History> = historyInitialiser(stockName, from, to)
            val retAlgorithm: Algorithm = when (type) {
                Type.TACPP46 -> TACPP46(init, retHistory.toMutableList())
            }
            return Pair(retHistory, retAlgorithm)
        }

        //===========================================================//
        /**
         * Loads historical data from files based on the given parameters.
         *
         * @param stockName - Stock identifier/name.
         * @param from     - Start date (inclusive).
         * @param to       - End date (inclusive).
         * @return         List of historical data entries.
         */
        private fun historyInitialiser(stockName: String, from: Int, to: Int): List<History> {
            val backtestFiles = File("src/main/resources/backtest/us/").listFiles() ?: error("Backtest directory is missing or not a directory")
            val proxy: MutableList<Pair<File, Int>> = ArrayList()

            for (file in backtestFiles) {
                val nameWithoutExtension = file.nameWithoutExtension
                val date = Integer.parseInt(nameWithoutExtension.substring(nameWithoutExtension.length - 2))
                val nameWithoutExtensionAndDate = nameWithoutExtension.substring(0, nameWithoutExtension.length - 2)

                if (date in from..to && nameWithoutExtensionAndDate == stockName) proxy.add(Pair(file, date))
            }


            proxy.sortWith(Comparator.comparingInt { h -> h.second }) // Sort chronologically by the date
            // If 'from' and/or 'to' are MIN and/or MAX then we've loaded all, so we don't throw.
            require(!(from != Integer.MIN_VALUE && to != Integer.MAX_VALUE && proxy.size != (to - from + 1))) { "From-To" }

            val ret: MutableList<History> = ArrayList()
            for (item in proxy) {
                val serData = SerializationManager.loadFromFile(item.first)
                for (item2 in serData.stockHistory.entries) {
                    ret.addAll(item2.value)
                }
            }

            return ret
        }
    }

    //===========================================================//
    //===========================================================//
    // Enum(s)

    /**
     * Supported algorithm types.
     */
    enum class Type {
        TACPP46,
    }

    //===========================================================//
    /**
     * Initialization modes for the algorithm.
     */
    internal enum class Init {
        BACKTEST,
        TRADING
    }

    //===========================================================//
    //===========================================================//
    // Helper class(es)

    data class Output(
        val buy: Buy?,
        val sell: Sell?
    ) {
        data class Buy(val amount: Long)
        data class Sell(val batches: List<Pair<Holding, Long>>)
    }

    //===========================================================//
    //===========================================================//
    // Constructor(s)

    /**
     * Package constructor to enforce initialization mode handling
     * in subclasses.
     *
     * @param init - Initialization mode.
     */
    internal constructor(init: Init)
}